// @myriaddreamin/typst.ts/dist/cjs/contrib/dom/typst-cancel.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypstCancellationToken = void 0;
class TypstCancellationToken {
    isCancellationRequested = false;
    _onCancelled;
    _onCancelledResolveResolved;
    constructor() {
        let resolveT = undefined;
        let resolveX = undefined;
        this._onCancelled = new Promise(resolve => {
            resolveT = resolve;
            if (resolveX) {
                resolveX(resolve);
            }
        });
        this._onCancelledResolveResolved = new Promise(resolve => {
            resolveX = resolve;
            if (resolveT) {
                resolve(resolveT);
            }
        });
    }
    async cancel() {
        await this._onCancelledResolveResolved;
        this.isCancellationRequested = true;
    }
    isCancelRequested() {
        return this.isCancellationRequested;
    }
    async consume() {
        (await this._onCancelledResolveResolved)();
    }
    wait() {
        return this._onCancelled;
    }
}
exports.TypstCancellationToken = TypstCancellationToken;
