// @myriaddreamin/typst.ts/dist/cjs/fs/memory.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryAccessModel = void 0;
class MemoryAccessModel {
    mTimes = new Map();
    mData = new Map();
    constructor() { }
    reset() {
        this.mTimes.clear();
        this.mData.clear();
    }
    insertFile(path, data, mtime) {
        this.mTimes.set(path, mtime);
        this.mData.set(path, data);
    }
    removeFile(path) {
        this.mTimes.delete(path);
        this.mData.delete(path);
    }
    getMTime(path) {
        if (!path.startsWith('/@memory/')) {
            return undefined;
        }
        if (this.mTimes.has(path)) {
            return this.mTimes.get(path);
        }
        return undefined;
    }
    isFile() {
        return true;
    }
    getRealPath(path) {
        return path;
    }
    readAll(path) {
        if (!path.startsWith('/@memory/')) {
            return undefined;
        }
        if (this.mData.has(path)) {
            return this.mData.get(path);
        }
        return undefined;
    }
}
exports.MemoryAccessModel = MemoryAccessModel;
