// @myriaddreamin/typst.ts/dist/cjs/main.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTypstCompiler = exports.FetchPackageRegistry = exports.FetchAccessModel = exports.createTypstSvgRenderer = exports.createTypstRenderer = exports.rendererBuildInfo = exports.preloadSystemFonts = exports.preloadRemoteFonts = void 0;
const initOptions = require("./options.init.cjs");
var options_init_mjs_1 = require("./options.init.cjs");
Object.defineProperty(exports, "preloadRemoteFonts", { enumerable: true, get: function () { return options_init_mjs_1.preloadRemoteFonts; } });
Object.defineProperty(exports, "preloadSystemFonts", { enumerable: true, get: function () { return options_init_mjs_1.preloadSystemFonts; } });
const renderer = require("./renderer.cjs");
var renderer_mjs_1 = require("./renderer.cjs");
Object.defineProperty(exports, "rendererBuildInfo", { enumerable: true, get: function () { return renderer_mjs_1.rendererBuildInfo; } });
Object.defineProperty(exports, "createTypstRenderer", { enumerable: true, get: function () { return renderer_mjs_1.createTypstRenderer; } });
Object.defineProperty(exports, "createTypstSvgRenderer", { enumerable: true, get: function () { return renderer_mjs_1.createTypstSvgRenderer; } });
const view_mjs_1 = require("./render/canvas/view.cjs");
const compiler = require("./compiler.cjs");
const index_mjs_1 = require("./fs/index.cjs");
const package_mjs_1 = require("./fs/package.cjs");
var index_mjs_2 = require("./fs/index.cjs");
Object.defineProperty(exports, "FetchAccessModel", { enumerable: true, get: function () { return index_mjs_2.FetchAccessModel; } });
var package_mjs_2 = require("./fs/package.cjs");
Object.defineProperty(exports, "FetchPackageRegistry", { enumerable: true, get: function () { return package_mjs_2.FetchPackageRegistry; } });
var compiler_mjs_1 = require("./compiler.cjs");
Object.defineProperty(exports, "createTypstCompiler", { enumerable: true, get: function () { return compiler_mjs_1.createTypstCompiler; } });
// Export module on window.
if (typeof window !== 'undefined') {
    window.TypstRenderModule = {
        RenderView: view_mjs_1.RenderView,
        createTypstRenderer: renderer.createTypstRenderer,
        createTypstSvgRenderer: renderer.createTypstRenderer,
        preloadRemoteFonts: initOptions.preloadRemoteFonts,
        preloadSystemFonts: initOptions.preloadSystemFonts,
    };
    window.TypstCompileModule = {
        createTypstCompiler: compiler.createTypstCompiler,
        preloadRemoteFonts: initOptions.preloadRemoteFonts,
        preloadSystemFonts: initOptions.preloadSystemFonts,
        FetchAccessModel: index_mjs_1.FetchAccessModel,
        MemoryAccessModel: index_mjs_1.MemoryAccessModel,
        FetchPackageRegistry: package_mjs_1.FetchPackageRegistry,
        withAccessModel: initOptions.withAccessModel,
        withPackageRegistry: initOptions.withPackageRegistry,
    };
}
